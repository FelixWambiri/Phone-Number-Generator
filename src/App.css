/*----------------------------*/
/*BASIC SETUP*/
/*----------------------------*/
* {
  box-sizing: border-box;
  -moz-box-sizing: border-box;
  -webkit-box-sizing: border-box;
  margin: 0;
  padding: 0;
}

/*----------------------------*/
/* Main Application Container */
/*----------------------------*/
.App {
  text-align: center;
  min-height: 100vh;
  background-image: linear-gradient(rgba(0, 0, 0, 0.6), rgba(0, 0, 0, 0.6)),
    url(images/tech.jpg);
  background-size: cover;
  background-position: center;
}

/*-----------------------------*/
/* Right Section of The Header */
/* Download, Ascending and     */
/* Descending Buttons          */
/*-----------------------------*/
.right-header-content {
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
  margin-top: 50px;
  position: absolute;
  right: 10%;
}

/*-----------------------------*/
/* Center Section of The Header*/
/* Maxmimum and Minimum number */
/* Buttons                     */
/*-----------------------------*/
.center-header-content {
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
  margin-top: 50px;
  position: absolute;
  left: 40%;
}

/*-----------------------------*/
/* Common Styling for all the  */
/* Icon Buttons                */
/*-----------------------------*/
.icon-btn-container {
  width: 60px;
  height: 60px;
  color: #e67e22;
  border-radius: 50%;
  background: #fff;
  padding-top: 5px;
  margin-left: 50px;
  transition: background-color 0.3s;
}

/*-----------------------------*/
/* Styling for the first Icon  */
/* Button in center header     */
/* container                   */
/*-----------------------------*/
.center-header-content .icon-btn-container:nth-child(1) {
  margin-left: 0;
  margin-right: 95px;
}

/*-----------------------------*/
/* Common Styling for all the  */
/* Icon Buttons on Hover       */
/*-----------------------------*/
.icon-btn-container:hover,.icon-btn-container:active {
  background-color: #d3d3d3;
  cursor: pointer !important;
}

/*-----------------------------*/
/* Common Styling for all the  */
/* Icons on Hover              */
/*-----------------------------*/
.icon-btn-container:hover svg path {
  fill: #cf6d17;
}
.icon-btn-container:active svg path {
  fill: #cf6d17;
}

/*-----------------------------*/
/* Common Styling for all the  */
/* Sections                    */
/*-----------------------------*/
section {
  position: absolute;
  top: 20%;
  left: 3%;
}

.left-section {
  width: 28vw;
  height: 50vh;
}

/*-----------------------------*/
/* Styling for all the right   */
/* Section                     */
/*-----------------------------*/
.right-section {
  background-image: url(images/white.jpg);
  background-size: cover;
  background-position: center;
  height: 68vh;
  width: 28vw;
  left: 67%;
  top: 20%;
  box-shadow: inset 0 0 15px rgba(0, 0, 0, 0.5);
  overflow-y: scroll;
  color: #000;
}

/*--------------------------------------------------------*/
/*Start of Styling for the list displaying numbers*/
/*--------------------------------------------------------*/
ul {
  margin: 0 auto;
  box-sizing: border-box;
  background: rgba(0, 0, 0, 0.1);
  border-radius: 5px;
}

ul li {
  display: flex;
  background: rgba(16, 16, 16, 0.9);
  padding: 10px 20px;
  line-height: 1.5rem;
  border: 1px solid #e0e0e0;
  color: #fff;
  margin: 5px 0;
  transition: all 0.3s linear;
}

ul li:nth-child(even) {
  background: rgba(19, 17, 17, 0.8);
}

ul li:hover {
  transform: scale(1.06);
  background: #ff3772;
  box-shadow: 0 14px 28px rgba(0, 0, 0, 0.25), 0 10px 10px rgba(0, 0, 0, 0.22);
  transform: scale(1.05);
}
/*--------------------------------------------------------*/
/*End of Styling for the list displaying numbers*/
/*--------------------------------------------------------*/

/*-----------------------------*/
/* Styling for all the middle  */
/* Section                     */
/*-----------------------------*/
.middle-section {
  top: 20%;
  left: 33%;
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
}

/*------------------------------*/
/* Styling for all the maxmimum */
/* and minimun number containers*/
/* in the middleSection         */
/*------------------------------*/
.card-container {
  width: 15vw;
  height: 15vh;
  background: rgba(255, 255, 255, 0.9);
  margin-left: 10px;
  font-size: 25px;
  letter-spacing: 1px;
  text-align: center;
  transition: 0.5s;
}

/*------------------------------*/
/* Styling for all the maxmimum */
/* and minimun number containers*/
/* in the middleSection on Hover*/
/*------------------------------*/
.card-container:hover, .card-container:active {
  transform: scale(1.06);
  background: #ff3772;
}

/*--------------------------------------------------------*/
/*Media Queries for different screen sizes*/
/*--------------------------------------------------------*/

/*
  ##Device = Desktops
  ##Screen = 1600px to higher resolution desktops
*/
@media (min-width: 1600px) {
  .center-header-content {
    left: 40%;
  }
  .middle-section {
    left: 32%;
  }
  .center-header-content .icon-btn-container:nth-child(1) {
    margin-left: 0;
    margin-right: 215px;
  }
  .card-container {
    margin-left: 30px;
    margin-right: 40px;
    font-size: 30px;
  }
  .right-section {
    left: 70%;
  }
}

/*
  ##Device = Tablets, Ipads (portrait)
  ##Screen = B/w 768px to 1024px
*/
@media only screen and (min-device-width: 768px) and (max-device-width: 1024px) and (orientation: portrait) and (-webkit-min-device-pixel-ratio: 1) {
  .App {
    background-image: linear-gradient(rgba(0, 0, 0, 0.1), rgba(0, 0, 0, 0.1)),
      url(images/tech.jpg);
  }
  .center-header-content {
    left: 10%;
  }
  .center-header-content .icon-btn-container:nth-child(1) {
    margin-right: 0;
  }
  .right-header-content {
    right: 10%;
  }
  .icon-btn-container {
    width: 100px;
    height: 100px;
    margin-left: 75px;
  }
  .middle-section {
    top: 14%;
    left: 5%;
  }
  .card-container {
    width: 42vw;
    height: 7vh;
    margin-left: 17px;
    font-size: 27px;
  }
  .right-section {
    top: 23%;
    left: 50%;
    width: 45vw;
    font-size: 30px;
  }
  .left-section {
    top: 23%;
    width: 45vw;
    left: 3%;
  }
  .left-section div {
    font-size: 30px !important;
  }
  .left-section input {
    color: rgb(238, 156, 52) !important;
    font-size: 30px !important;
    height: 40px !important;
    margin-top: 0 !important;
    padding: 0 !important;
  }
}

/*
  ##Device = Most of the Smartphones Mobiles (Portrait)
  ##Screen = B/w 320px to 479px
*/
@media (min-width: 320px) and (max-width: 480px) {
  .App {
    overflow-y: scroll !important;
  }
  .center-header-content .icon-btn-container:nth-child(1) {
    margin-left: 0;
    margin-right: 0;
  }
  .icon-btn-container {
    margin-left: 10px;
  }
  .right-header-content {
    right: 5%;
  }
  .center-header-content {
    left: 5%;
  }
  .card-container {
    width: 85vw;
    height: 10vh;
    margin-bottom: 5px;
    font-size: 15px;
  }
  .middle-section {
    top: 18%;
    left: 5%;
    flex-direction: column;
  }
  .right-section {
    top: 90%;
    left: 3%;
    width: 95vw;
  }
  .left-section {
    top: 40%;
    width: 95vw;
    height: 15vh;
  }
}
