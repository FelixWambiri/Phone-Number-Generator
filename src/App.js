import React from "react";

// Third Party libraries
import { ic_file_download } from "react-icons-kit/md/ic_file_download";
import { ic_arrow_downward } from "react-icons-kit/md/ic_arrow_downward";
import { ic_arrow_upward } from "react-icons-kit/md/ic_arrow_upward";
import { ic_vertical_align_bottom } from "react-icons-kit/md/ic_vertical_align_bottom";
import { ic_vertical_align_top } from "react-icons-kit/md/ic_vertical_align_top";
import Workbook from "react-excel-workbook";

// Components
import AppTerminal from "./components/Terminal";
import Button from "./common/Button";
import NumberList from "./components/NumberList";

// Helper functions
import { generateRandom, sortArrays, findMax, findMin } from "./utils";

// Styling
import "./App.css";
class App extends React.Component {
  state = {
    phoneNumbers: [],
    maxNumber: "",
    minNumber: ""
  };

  getLocalStorageItem = name => localStorage.getItem(name);
  setLocalStorageItem = (key, value) =>
    localStorage.setItem(`${key}`, JSON.stringify(value));

  componentDidMount() {
    const phoneNumbers = this.getLocalStorageItem("phoneNumbers")
      ? JSON.parse(this.getLocalStorageItem("phoneNumbers"))
      : [];

    const keysToGet = ["maxNumber", "minNumber"];
    const localValues = keysToGet.map(key => {
      if (this.getLocalStorageItem(key)) {
        return JSON.parse(this.getLocalStorageItem(key));
      } else {
        return "";
      }
    });

    const [maxNumber, minNumber] = localValues;
    if (phoneNumbers || maxNumber || minNumber) {
      this.setState({
        phoneNumbers,
        maxNumber,
        minNumber
      });
    }
  }

  componentDidUpdate(prevProps, prevState) {
    const { phoneNumbers, maxNumber, minNumber } = this.state;

    if (phoneNumbers !== prevState.phoneNumbers) {
      this.setLocalStorageItem("phoneNumbers", phoneNumbers);
    }
    if (maxNumber !== prevState.maxNumber) {
      this.setLocalStorageItem("maxNumber", maxNumber);
    }
    if (minNumber !== prevState.minNumber) {
      this.setLocalStorageItem("minNumber", minNumber);
    }
  }

  generateRandomPhoneNumbers = value => {
    try {
      const phoneNumbers = generateRandom(value);
      this.setState({
        phoneNumbers
      });
      return `Generating Phone Numbers...`;
    } catch (error) {
      console.log("The error is ", error);
    }
  };

  sortPhoneNumbers = (array, order) => {
    const phoneNumbers = sortArrays(array, order);
    this.setState({
      phoneNumbers
    });
  };

  getMaxNumber = arr => {
    const maxNumber = findMax(arr);
    this.setState({
      maxNumber
    });
  };

  getMinNumber = arr => {
    const minNumber = findMin(arr);
    this.setState({
      minNumber
    });
  };

  deleteGeneratedNumbers = () => {
    let keysToRemove = ["phoneNumbers", "maxNumber", "minNumber"];
    keysToRemove.forEach(key => localStorage.removeItem(key));
    this.setState({
      phoneNumbers: [],
      maxNumber: "",
      minNumber: ""
    });
  };

  render() {
    const { phoneNumbers, maxNumber, minNumber } = this.state;
    const data = phoneNumbers.map(num => ({ phone: num }));
    return (
      <div className="App">
        <header>
          <div className="right-header-content">
            <Workbook
              filename="phone-numbers.xlsx"
              element={
                <Button
                  title="Download Numbers Generated"
                  placement="top"
                  iconClassName="icon-btn-container"
                  size={"80%"}
                  icon={ic_file_download}
                />
              }
            >
              <Workbook.Sheet data={data} name="Phone Numbers">
                <Workbook.Column label="Phone Numbers" value="phone" />
              </Workbook.Sheet>
            </Workbook>

            <Button
              title="Ascending Order"
              placement="top"
              iconClassName="icon-btn-container"
              size={"80%"}
              icon={ic_arrow_upward}
              onClick={() => this.sortPhoneNumbers(phoneNumbers, "ascending")}
            />

            <Button
              title="Descending Order"
              placement="top"
              iconClassName="icon-btn-container"
              size={"80%"}
              icon={ic_arrow_downward}
              onClick={() => this.sortPhoneNumbers(phoneNumbers, "descending")}
            />
          </div>

          <div className="center-header-content">
            <Button
              title="Maximum Number"
              placement="top"
              iconClassName="icon-btn-container"
              size={"80%"}
              icon={ic_vertical_align_top}
              onClick={() => this.getMaxNumber(phoneNumbers)}
            />
            <Button
              title="Minimum Number"
              placement="top"
              iconClassName="icon-btn-container"
              size={"80%"}
              icon={ic_vertical_align_bottom}
              onClick={() => this.getMinNumber(phoneNumbers)}
            />
          </div>
        </header>
        <section className="left-section">
          <AppTerminal
            generatePhoneNumbers={this.generateRandomPhoneNumbers}
            deleteGeneratedNumbers={this.deleteGeneratedNumbers}
          />
        </section>
        <section className="middle-section">
          {maxNumber && (
            <div className="card-container">
              <p>Biggest Number</p>
              <br />
              <p>0{maxNumber}</p>
            </div>
          )}
          {minNumber && (
            <div className="card-container">
              <p>Smallest Number</p>
              <br />
              <p>0{minNumber}</p>
            </div>
          )}
        </section>
        <section className="right-section">
          <ul>
            {phoneNumbers.length > 0 &&
              phoneNumbers.map((num, index) => {
                return <NumberList phoneNumber={num} key={index} />;
              })}
          </ul>
        </section>
      </div>
    );
  }
}

export default App;
