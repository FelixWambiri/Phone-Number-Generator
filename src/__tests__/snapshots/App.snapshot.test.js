import React from "react";
import renderer from "react-test-renderer";

// Test Component
import App from "../../App";

it("Should render with all the child components being empty", () => {
  const tree = renderer.create(<App />);

  expect(tree).toMatchSnapshot();
});
