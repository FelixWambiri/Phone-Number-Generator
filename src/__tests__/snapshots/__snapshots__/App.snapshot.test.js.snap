// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render with all the child components being empty 1`] = `
<div
  className="App"
>
  <header>
    <div
      className="right-header-content"
    >
      <span
        onClick={[Function]}
      >
        <div
          aria-describedby={null}
          className="icon-btn-container"
          onBlur={[Function]}
          onFocus={[Function]}
          onMouseLeave={[Function]}
          onMouseOver={[Function]}
          onTouchEnd={[Function]}
          onTouchStart={[Function]}
          title="Download Numbers Generated"
        >
          <i
            fill="currentColor"
            style={
              Object {
                "display": "inline-block",
              }
            }
          >
            <svg
              fill="currentColor"
              height="80%"
              style={
                Object {
                  "display": "inline-block",
                  "verticalAlign": "middle",
                }
              }
              viewBox="0 0 24 24"
              width="80%"
            >
              <path
                d="M19 9h-4V3H9v6H5l7 7 7-7zM5 18v2h14v-2H5z"
              />
            </svg>
          </i>
        </div>
      </span>
      <div
        aria-describedby={null}
        className="icon-btn-container"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        title="Ascending Order"
      >
        <i
          fill="currentColor"
          style={
            Object {
              "display": "inline-block",
            }
          }
        >
          <svg
            fill="currentColor"
            height="80%"
            style={
              Object {
                "display": "inline-block",
                "verticalAlign": "middle",
              }
            }
            viewBox="0 0 24 24"
            width="80%"
          >
            <path
              d="M4 12l1.41 1.41L11 7.83V20h2V7.83l5.58 5.59L20 12l-8-8-8 8z"
            />
          </svg>
        </i>
      </div>
      <div
        aria-describedby={null}
        className="icon-btn-container"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        title="Descending Order"
      >
        <i
          fill="currentColor"
          style={
            Object {
              "display": "inline-block",
            }
          }
        >
          <svg
            fill="currentColor"
            height="80%"
            style={
              Object {
                "display": "inline-block",
                "verticalAlign": "middle",
              }
            }
            viewBox="0 0 24 24"
            width="80%"
          >
            <path
              d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
            />
          </svg>
        </i>
      </div>
    </div>
    <div
      className="center-header-content"
    >
      <div
        aria-describedby={null}
        className="icon-btn-container"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        title="Maximum Number"
      >
        <i
          fill="currentColor"
          style={
            Object {
              "display": "inline-block",
            }
          }
        >
          <svg
            fill="currentColor"
            height="80%"
            style={
              Object {
                "display": "inline-block",
                "verticalAlign": "middle",
              }
            }
            viewBox="0 0 24 24"
            width="80%"
          >
            <path
              d="M8 11h3v10h2V11h3l-4-4-4 4zM4 3v2h16V3H4z"
            />
          </svg>
        </i>
      </div>
      <div
        aria-describedby={null}
        className="icon-btn-container"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        title="Minimum Number"
      >
        <i
          fill="currentColor"
          style={
            Object {
              "display": "inline-block",
            }
          }
        >
          <svg
            fill="currentColor"
            height="80%"
            style={
              Object {
                "display": "inline-block",
                "verticalAlign": "middle",
              }
            }
            viewBox="0 0 24 24"
            width="80%"
          >
            <path
              d="M16 13h-3V3h-2v10H8l4 4 4-4zM4 19v2h16v-2H4z"
            />
          </svg>
        </i>
      </div>
    </div>
  </header>
  <section
    className="left-section"
  >
    <div
      className="terminal"
      name="react-console-emulator"
      onClick={[Function]}
      style={
        Object {
          "background": "linear-gradient(rgba(102,153,102,0.2), rgba(255, 255, 255, 0.3)),url(test-file-stub)",
          "backgroundSize": "cover",
          "borderRadius": "5px",
          "cursor": "text",
          "height": "100%",
          "maxHeight": "100%",
          "maxWidth": "100%",
          "minHeight": "300px",
          "minWidth": "500x",
          "overflow": "auto",
          "textAlign": "left",
          "width": "100%",
        }
      }
    >
      <div
        name="react-console-emulator__content"
        style={
          Object {
            "color": "#FFFFFF",
            "fontFamily": "monospace",
            "fontSize": "15px",
            "height": "100%",
            "padding": "20px",
          }
        }
      >
        <p
          style={
            Object {
              "lineHeight": "21px",
              "margin": "0",
            }
          }
        >
          Welcome... :)
        </p>
        <p
          style={
            Object {
              "lineHeight": "21px",
              "margin": "0",
            }
          }
        >
          Type help to view a list of available commands
        </p>
        <div
          name="react-console-emulator__inputArea"
          style={
            Object {
              "display": "inline-flex",
              "width": "100%",
            }
          }
        >
          <span
            name="react-console-emulator__promptLabel"
            style={
              Object {
                "color": "#EE9C34",
                "paddingTop": "3px",
              }
            }
          >
            @phoneNumbers:~$
          </span>
          <input
            autoComplete="off"
            disabled={false}
            name="react-console-emulator__input"
            onKeyDown={[Function]}
            style={
              Object {
                "background": "transparent",
                "border": "0",
                "color": "#F0BF81",
                "flexGrow": "100",
                "fontFamily": "monospace",
                "fontSize": "15px",
                "height": "22px",
                "margin": "0",
                "outline": "none",
                "padding": "0 0 0 7px",
                "width": "100%",
              }
            }
            type="text"
          />
        </div>
      </div>
    </div>
  </section>
  <section
    className="middle-section"
  >
    
    
  </section>
  <section
    className="right-section"
  >
    <ul />
  </section>
</div>
`;
